collector/infiniband_linux.go

```diff
@@ -0,0 +1,167 @@
+		"unicast_rcv_packets":    "Number of unicast packets received (including errors)",
+		"unicast_xmit_packets":   "Number of unicast packets transmitted (including errors)",
+	}
+
+	subsystem := "infiniband"
+	i.metricDescs = make(map[string]*prometheus.Desc)
+
+	for metric, description := range i.counters {
+		i.metricDescs[metric] = prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, metric+"_total"),
+			description,
+			[]string{"device", "port"},
+			nil)
+	}
+
+	return &i, nil
+}
+
+// Retrieve a list of InfiniBand devices.
+func infinibandDevices(infinibandPath string) ([]string, error) {
+	devices, err := filepath.Glob(filepath.Join(infinibandPath, "/*"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(devices) < 1 {
+		log.Debugf("Unable to detect InfiniBand devices")
+		err = infinibandNoDevicesFound
+		return nil, err
+	}
+
+	// Extract just the filenames which equate to the device names.
+	for i, device := range devices {
+		devices[i] = filepath.Base(device)
+	}
+
+	return devices, nil
+}
+
+// Retrieve a list of ports for the InfiniBand device.
+func infinibandPorts(infinibandPath, device string) ([]string, error) {
+	ports, err := filepath.Glob(filepath.Join(infinibandPath, device, "ports/*"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ports) < 1 {
+		log.Debugf("Unable to detect ports for %s", device)
+		err = infinibandNoPortsFound
+		return nil, err
+	}
+
+	// Extract just the filenames which equates to the port numbers.
+	for i, port := range ports {
+		ports[i] = filepath.Base(port)
+	}
+
+	return ports, nil
+}
+
+func readMetric(directory, metricFile string) (uint64, error) {
+	metric, err := readUintFromFile(filepath.Join(directory, metricFile))
+
+	if err != nil {
+		log.Debugf("Error reading " + metricFile + " file")
+		return 0, err
+	}
+
+	return metric, nil
+}
+
+func (c *infinibandCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	devices, err := infinibandDevices(sysFilePath(infinibandPath))
+
+	// If no devices are found or another error is raised while attempting to find devices,
+	// InfiniBand is likely not installed and the collector should be skipped.
+	if err == infinibandNoDevicesFound || err != nil {
+		return nil
+	}
+
+	for _, device := range devices {
+		ports, err := infinibandPorts(sysFilePath(infinibandPath), device)
+
+		// If no ports are found for the specified device, skip to the next device.
+		if err == infinibandNoPortsFound {
+			continue
+		} else if err != nil {
+			return nil
+		}
+
+		for _, port := range ports {
+			portFiles := sysFilePath(filepath.Join(infinibandPath, device, "ports", port))
+
+			// Add metrics for the InfiniBand counters.
+			for counterFile := range c.counters {
+				metric, err := readMetric(filepath.Join(portFiles, "counters"), counterFile)
+
+				if err != nil {
+					return err
+				}
+
+				ch <- prometheus.MustNewConstMetric(c.metricDescs[counterFile],
```

Comment:
Same as previous nit: give each argument its own line and have the parenthesis line up at the end.